gson/src/main/java/com/google/gson/internal/bind/TypeAdapterRuntimeTypeWrapper.java

```diff
@@ -51,7 +52,7 @@ public void write(JsonWriter out, T value) throws IOException {
 
     TypeAdapter chosen = delegate;
     Type runtimeType = getRuntimeTypeIfMoreSpecific(type, value);
-    if (runtimeType != type) {
+    if (runtimeType != type && !Primitives.areEquivalentTypes(runtimeType, type)) {

```

Comment:
This doesn’t seem like the right location for the fix. I’d expect it to be nearby the `@JsonAdapter` handling code.